### PR TITLE
[FEATURE] Feature/zero product

### PR DESCRIPTION
**Pull request by trickreich**

### CODE REVIEW

**Comment by christianbader** [comment]
tests?

File: Api/Product.php [hunk 1]
```diff
@@ -1228,4 +1228,9 @@ public function getSearchTerms()
     {
         return $this->entity->getSearchTerms();
     }
+
+    public function isValidShopProduct($defaultCurrency)
```
Comment (christianbader) [inline Api/Product.php]: add phpdoc

File: Entity/BaseProduct.php [hunk 2]
```diff
@@ -1045,18 +1045,43 @@ public function setSearchTerms($searchTerms)
     /**
      * Helper method to check if the product is
      * a valid shop product.
-     *
+     * @param $defaultCurrency
      * @return bool
      */
-    public function isValidShopProduct()
+    public function isValidShopProduct($defaultCurrency)
```
Comment (christianbader) [inline Entity/BaseProduct.php]: declare a variable isValid = false;

File: Entity/BaseProduct.php [hunk 3]
```diff
@@ -1045,18 +1045,43 @@ public function setSearchTerms($searchTerms)
...
     {
         if (method_exists($this, 'getPrices') &&
             $this->getStatus()->getId() == Status::ACTIVE &&
             $this->getPrices() && count($this->getPrices()) > 0
         ) {
+            if(!$this->hasPriceInDefaultCurrency($this->getPrices(), $defaultCurrency)) {
```
Comment (christianbader) [inline Entity/BaseProduct.php]: +            if($this->hasPriceInDefaultCurrency($this->getPrices(), $defaultCurrency)) {
 +                $isValid = true;
 +            }
 +            if($this->getSupplier()) {
 +                $isValid = true;
 +            }
 +          }
  
          return $isValid;
      }

File: Entity/BaseProduct.php [hunk 4]
```diff
@@ -1045,18 +1045,43 @@ public function setSearchTerms($searchTerms)
...
 
         return false;
     }
+
+    /**
+     * Checks if price in default currency exists
+     *
+     * @param ProductPrice[] $prices
+     * @param $defaultCurrency
+     * @return boolean
```
Comment (christianbader) [inline Entity/BaseProduct.php]: always add an empty line before return

File: Entity/ProductInterface.php [hunk 5]
```diff
@@ -528,4 +528,13 @@ public function setSearchTerms($searchTerms);
      * @return string
      */
     public function getSearchTerms();
+
+    /**
+     * Helper method to check if the product is
+     * a valid shop product.
+     *
+     * @param $defaultCurrency
+     * @return bool
```
Comment (christianbader) [inline Entity/ProductInterface.php]: empty line before return

File: Product/ProductManager.php [hunk 6]
```diff
@@ -12,6 +12,8 @@
 
 use DateTime;
 use Doctrine\Common\Persistence\ObjectManager;
+use Sulu\Bundle\ProductBundle\Product\Exception\ProductException;
```
Comment (christianbader) [inline Product/ProductManager.php]: order! move down

File: Product/ProductManager.php [hunk 7]
```diff
@@ -1331,6 +1333,14 @@ public function save(
             $product->setDeliveryStatus($deliveryStatus);
         }
 
+        if($product->getStatus()->getId() == StatusEntity::ACTIVE) {
+            // if the status of the product is active then the product must be a valid shop product!
```
Comment (christianbader) [inline Product/ProductManager.php]: If the status ... Uppercase ;)

File: Product/ProductManager.php [hunk 8]
```diff
@@ -1654,7 +1665,7 @@ public function setDeliveryStatusForProduct($product, $statusId)
      *
      * @param ProductPrice $price
      * @param array $matchedEntry
-     * @throws Exception\ProductDependencyNotFoundException
+     * @throws ProductDependencyNotFoundException
      * @return bool
```
Comment (christianbader) [inline Product/ProductManager.php]: empty line before return -> check whole file please.

File: Resources/public/js/components/products/main.js [hunk 9]
```diff
@@ -140,6 +141,28 @@ define([
             this.sandbox.on('sulu.header.back', function() {
                 this.sandbox.emit('sulu.products.list');
             }, this);
+
+            // products save-error
```
Comment (christianbader) [inline Resources/public/js/components/products/main.js]: Usse more meaningful comments.

File: Resources/public/js/components/products/main.js [hunk 10]
```diff
@@ -140,6 +141,28 @@ define([
...
+                            'sulu.labels.error.show',
+                            'labels.error.product-could-not-be-activated',
+                            'labels.error'
+                        );
+
+                        this.status = Config.get('product.status.inactive');
+                        this.sandbox.emit('product.state.change', this.status);
+                    }
+                    else {
+                        // code not defined
```
Comment (christianbader) [inline Resources/public/js/components/products/main.js]: same here

File: Api/Product.php [hunk 11]
```diff
@@ -1228,4 +1228,17 @@ public function getSearchTerms()
     {
         return $this->entity->getSearchTerms();
     }
+
+    /**
+     * Helper method to check if the product is
+     * a valid shop product.
+     *
+     * @param $defaultCurrency
```
Comment (turbo-ele) [inline Api/Product.php]: string

File: Entity/BaseProduct.php [hunk 12]
```diff
@@ -1046,15 +1077,41 @@ public function setSearchTerms($searchTerms)
      * Helper method to check if the product is
      * a valid shop product.
      *
+     * @param $defaultCurrency
```
Comment (turbo-ele) [inline Entity/BaseProduct.php]: string

File: Entity/BaseProduct.php [hunk 13]
```diff
@@ -1046,15 +1077,41 @@ public function setSearchTerms($searchTerms)
...
+        }
+
+        return $isValid;
+    }
+
+    /**
+     * Checks if price in default currency exists
+     *
+     * @param ProductPrice[] $prices
+     * @param $defaultCurrency
```
Comment (turbo-ele) [inline Entity/BaseProduct.php]: string

File: Entity/BaseProduct.php [hunk 14]
```diff
@@ -1046,15 +1077,41 @@ public function setSearchTerms($searchTerms)
...
+     *
+     * @return boolean
```
Comment (turbo-ele) [inline Entity/BaseProduct.php]: bool

File: Entity/ProductInterface.php [hunk 15]
```diff
@@ -528,4 +528,14 @@ public function setSearchTerms($searchTerms);
      * @return string
      */
     public function getSearchTerms();
+
+    /**
+     * Helper method to check if the product is
+     * a valid shop product.
+     *
+     * @param $defaultCurrency
```
Comment (turbo-ele) [inline Entity/ProductInterface.php]: string

File: Product/ProductManager.php [hunk 16]
```diff
@@ -1331,6 +1333,14 @@ public function save(
             $product->setDeliveryStatus($deliveryStatus);
         }
 
+        if($product->getStatus()->getId() == StatusEntity::ACTIVE) {
+            // If the status of the product is active then the product must be a valid shop product!
+            if(!$product->isValidShopProduct($this->defaultCurrency)) {
+                $this->setStatusForProduct($product, StatusEntity::INACTIVE);
+                throw new ProductException('No valid product for shop!',1);
```
Comment (turbo-ele) [inline Product/ProductManager.php]: reformat

File: Resources/public/js/components/products/components/attributes/main.js [hunk 17]
```diff
@@ -35,7 +35,7 @@ define([
             }.bind(this));
 
             this.sandbox.on('product.state.change', function(status) {
-                if (!this.options.data.status || this.options.data.status.id !== status.id) {
+                if (!this.options.data || !this.options.data.attributes.status || this.options.data.attributes.status.id !== status.id) {
```
Comment (turbo-ele) [inline Resources/public/js/components/products/components/attributes/main.js]: wrap

Comment (turbo-ele) [inline Resources/public/js/components/products/components/attributes/main.js]: sure if that'S correct

File: Resources/public/js/components/products/components/detail-form/main.js [hunk 18]
```diff
@@ -52,9 +52,10 @@ define([
         },
 
         bindCustomEvents: function() {
-            this.sandbox.on('product.state.change', function(id) {
-                if (!this.options.data || !this.options.data.attributes.status || this.options.data.attributes.status.id !== id) {
-                    this.status = {id: id};
+            this.sandbox.on('product.state.change', function(status) {
+                if (!this.options.data || !this.options.data.attributes.status || this.options.data.attributes.status.id !== status.id) {
```
Comment (turbo-ele) [inline Resources/public/js/components/products/components/detail-form/main.js]: please check

File: Product/ProductManager.php (hunk 16 above)
Comment (turbo-ele) [inline Product/ProductManager.php]: implement const ProductException::PRODUCT_NOT_VALID

File: Resources/public/js/components/products/main.js [hunk 19]
```diff
@@ -140,6 +141,29 @@ define([
...
+                        this.sandbox.emit(
+                            'sulu.labels.error.show',
+                            'labels.error.product-could-not-be-activated',
+                            'labels.error'
+                        );
+
+                        this.status = Config.get('product.status.inactive');
+                        this.sandbox.emit('product.state.change', this.status);
+                    }
+                    else {
```
Comment (turbo-ele) [inline Resources/public/js/components/products/main.js]: reformat

**Comment by turbo-ele** [comment]
else ack

**Pull request merged**